guides/hack/11-built-in-types/73-noreturn.md

```diff
@@ -8,4 +8,19 @@ function something_went_wrong(): noreturn {
 }
 ```
 
-`exit` is an example of a library function with a `noreturn` type.
+`invariant_violation` is an example of a library function with a `noreturn` type.
+
+`noreturn` informs the typesystem that code execution can not continue past a certain line.
+In combination with a conditional, you can refine variables, since the typesystem will take note.
+This is actually how [invariant](../expressions-and-operators/invariant) is [implemented](/hack/reference/function/HH.invariant).
```

Comment:
Did you mean `invariant_violation`? (`invariant` can return.) I'm also not sure if this is strictly true, IIRC `invariant_violation` existed long before the `noreturn` type so might still have custom implementation in the typechecker...

Reply:
Rereading the paragraph, I understand the confusion. The `This` in "This is actually how [invariant](../expressions-and-operators/invariant) is [implemented](/hack/reference/function/HH.invariant) refers to the code sample below.
`invariant(X, Y)` is actually shorthand for
```HACK
if (!(X)) {
  invariant_violation(Y);
}
```

I wrote this with the intention of demonstrating that the concept of checking a condition before a noreturn is a very common idiom. I will either need to rephrase this or turn the `invariant_violation` name into a link to the invariant docs and remove the Sentence `This ... implemented`. I have not yet decided.